core/aws-core/src/main/java/software/amazon/awssdk/awscore/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.awscore.endpointdiscovery;
+
+import java.net.URI;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+public final class EndpointDiscoveryRefreshCache {
+
+    private static final Logger log = Logger.loggerFor(EndpointDiscoveryRefreshCache.class);
+
+    private final Map<String, EndpointDiscoveryEndpoint> cache = new ConcurrentHashMap<>();
+
+    private final EndpointDiscoveryCacheLoader client;
+
+    private EndpointDiscoveryRefreshCache(EndpointDiscoveryCacheLoader client) {
+        this.client = client;
+    }
+
+    public static EndpointDiscoveryRefreshCache create(EndpointDiscoveryCacheLoader client) {
+        return new EndpointDiscoveryRefreshCache(client);
+    }
+
+    /**
+     * Abstract method to be implemented by each service to handle retrieving
+     * endpoints from a cache. Each service must handle converting a request
+     * object into the relevant cache key.
+     *
+     * @param defaultEndpoint - The default endpoint for the service
+     * @return The endpoint to use for this request
+     */
+    public URI get(String accessKey, EndpointDiscoveryRequest request, URI defaultEndpoint) {
+        String key = accessKey;
+        if (request.cacheKey().isPresent()) {
+            key = key + ":" + request.cacheKey().get();
+        }
+
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        if (endpoint == null) {
+            if (request.required()) {
+                return cache.computeIfAbsent(key, k -> discoverEndpoint(request).join()).endpoint();
+            } else {
+                EndpointDiscoveryEndpoint tempEndpoint = EndpointDiscoveryEndpoint.builder()
```

Comment:
Don't quite follow this. So if it's not required we return the default endpoint which is valid for sixty seconds. After that time we fetch the discovered endpoint? Would it make more sense to just return the default endpoint and immediately kick off a request to populate the cache with the discovered endpoint?

Reply:
This is done for two reasons:

1. For new users, there is going to be a small period of creating their DNS entry.

2. In the case of short-lived functions where the client is not being re-used, this helps to prevent an unnecessary call.